bundler/spec/runtime/inline_spec.rb

```diff
@@ -44,7 +44,7 @@ def script(code, options = {})
     end
   end
 
-  it "requires the gems" do
+  it "requires the gems", bundler: "< 3" do
```

Comment:
We should fix these deprecations and run those specs on Bundler 3 too.

Reply:
To be able to remove `bundler: "< 3"` checks, I would have to change the specs from using `gemfile(true)` to using `gemfile`. Some of these specs would need to change from `gemfile(true, :ui => my_ui) do` to `gemfile(:ui => my_ui) do`. This means modifying the method signature from `gemfile(install = false, options = {}, &gemfile)` to `gemfile(install = false, **options, &gemfile)`, otherwise `install` will have the value `{ :ui => my_ui }` and `options` will be empty — which I guess we should fix anyway since that seems like an underlying bug.

Are we okay with me tinkering with the `gemfile` implementation in this PR?

Reply:
Also FYI because I would need to remove `true` from those tests' parameters, `install` will be false by default. Some of these tests like these ones:
    https://github.com/rubygems/rubygems/blob/4eb91f16b5f93ef2e486fc71bbe2f52e2257fec1/bundler/spec/runtime/inline_spec.rb#L103-L105

implicitly rely on install `true` to work correctly because they check output. If the install is not `true` then the current code implementation sets the ui to `silent` and this would break those specs:

https://github.com/rubygems/rubygems/blob/4eb91f16b5f93ef2e486fc71bbe2f52e2257fec1/bundler/lib/bundler/inline.rb#L38

The solution would be to either change `install`'s default value to `true` in the method definition, or change that line of code somehow like removing `|| !install` or other checks.

Reply:
@jeromedalbert Yes, we need to fix those specs, and the bug you found. Otherwise our own advice to fix the deprecation will not work. Also, we should probably explain the slight output change in the deprecation message and suggest an alternative that keeps the same behavior (with `quiet: true`).

I think we should keep the optional `install` parameter and handle the case where it's the only parameter and it's a `Hash` (meaning users have fixed the deprecation).

Reply:
@deivid-rodriguez I removed the `bundler: "< 3"` option and after some fixes I was able to make all tests pass but two: `inline_spec.rb:593` and `inline_spec.rb:689`. They fail because they used to do `gemfile(true)` which had the effect of always executing this block:

https://github.com/rubygems/rubygems/blob/465bfaf530c6712487495419e8ce9f5a0a0385e9/bundler/lib/bundler/inline.rb#L58-L65

But since I had to remove `(true)` to run the specs on Bundler 3, that block of code is not executed any more, and it seems like the test are set up incorrectly resulting in the definition not having missing specs, and the tests fail.

It seems to me that the tests are incorrect, rather than a regression in the code.  I don't understand enough about the setup part of these 2 tests to be able to rewrite them. Unless someone can find out what is wrong with these setups, I could either:

- Remove those 2 tests
- Use the `bundler: "< 3"` only for those 2 tests
- Remove the `if install || definition.missing_specs? { ... }` condition so we always attempt to install, which seems wrong as I don't think we should install unless specs are actually missing. I believe always installing would results in slow downs, and I do not think it is worth it just for the purpose of making two incorrect tests pass.

What are your thoughts?

Reply:
Thanks for the work, looking good!

Ok, so, so far we've discovered the actual differences between current `install = true` and `install = false`. With `install = true`, Bundler always has quiet output (even if explicit `quiet == false` is given? 🤨), and Bundler always force install gems (something like `bundle install --force`, which would be equivalent to a reinstall). With `install = false`, `quiet` option is respected and `false` by default, and gems are only installed when they are missing.

As per the remaining specs, `inline_spec.rb:593` is just not going that far now, but achieving the same result (no warnings printed), so the expectation on the error stream should be changed from "The Gemfile specifies no dependencies" to be empty.

Regarding `inline_spec.rb:689`, that's a funnier one. That spec was [introduced](https://github.com/rubygems/rubygems/pull/5535) when we added fallback behavior to `bundler/inline` so that if `bundler/inline` activates a default gem like `psych` internally, but then the inline Gemfile activates a different version of it, then we catch the error, add an explicit requirement on the version of psych activated internally (if compatible with the inline Gemfile), and retry, so that the error is avoided the second time. But now that we no longer force-install gems, `psych` is no longer activated and the spec no longer works as before.

So one way to change the spec to pass would be to change `psych` with some other gem that still gets activated internally, for example, `did_you_mean`. However, I tried changing it and noticed yet one more difference between `install = true` and `install = false`. With `install = true`, we'll always run a remote resolution and potentially install a higher versioned gem, even if a lower version that satisfies the Gemfile is present locally. So `install = true` is a way to always install the latest version regardless of what's present locally. And discovering this also gives some more sense to the (still confusing) current documentation of the `install` parameter, reading "whether gems that aren't already installed on the user's system should be installed".

Now that we know all this, I'm not so sure we want to deprecate this parameter. Maybe we should abort the deprecation, and instead:

* Rename the `install` parameter to something more descriptive like `force_latest`, maybe?
* Fix the bug with `quiet` being ignored when `install = true`.
* Improve the confusing documentation of the option mentioned in https://github.com/rubygems/rubygems/issues/5364.

Thoughts?

Reply:
Looks like we're getting somewhere! Regarding your 2 points (ignoring `quiet`):

- Something like `force_latest_compatible` makes sense (by the way would you want that param's default to remain `true` or changed to `false`?). The name is not as catchy and as punchy as the previous `install` parameter, but it's also more correct. So a `gemfile(force_latest_compatible = false, options = {}, &gemfile)` doesn't feel as direct. In fact, this force_latest compatible is kind of an oddly specific feature, which makes me wonder if it still deserves the prominence of being the first optional positional parameter. Almost makes me wonder if it should be relegated to an option in the `options = {}` hash. Again, not very knowledgeable about the codebase and this could have outsized impact, I'm just saying that from the point of view of someone who would be looking at this with new eyes I think `gemfile(options = {}, &gemfile)` may look better. On the other hand it might not be worth badgering people about deprecations just for arguably stylistic changes, and it might be better to let sleeping dogs lie and just rename the parameter without changing a single other thing.
- Yeah improving the confusing documentation makes sense for sure!

Reply:
Yeah I had the implication in mind, I don't know if it would be clear to others though, otherwise a fuller name like `force_latest_compatible` or `install_latest_compatible` but even then I think no matter what you'd need some good parameter description so mayb `install_latest` might be enough but I don't want to force my suggestion.

Reply:
I think any over the proposed alternative names is good enough, provided we write a clear description of the parameter. Are you up for repurposing this PR, and getting it over the finish line at this point?

Reply:
Sure if it's just a rename and some description I might open a new one